README.md
# Segment Plugin #

## Description ##

The plugin allows you to send a campaign to a segment, a subset, of the subscribers who belong to the selected lists.

The plugin adds a tab to the Send a campaign page that lets you define conditions. A condition comprises a field, an operator, and a target value.
When the campaign is sent only those subscribers who meet either any or all of the conditions will be selected.

These subscriber fields and attributes are supported for fields:

* each subscriber attribute
* the subscriber email address
* the subscriber id
* the subscriber unique id
* the date they signed-up to phpList
* the date they subscribed to the lists to which the campaign is being sent
* the subscriber belongs to all lists to which the campaign is being sent
* subscriber campaign activity - whether the subscriber was sent, opened, or clicked a link in a specific prior campaign

Each condition has a set of operators, specific to each field, such as 'is', 'is not', 'matches', 'is before', 'opened', etc.

The target value, also specific to each field, can be a text value, a select value, a date, etc.

## Installation ##

### Dependencies ###

Requires phplist version 3.3.2 or later, and php version 5.4.0 or later.
Please check your php version before installing the plugin, otherwise phplist will fail (probably a white page).

This plugin requires the Common Plugin version 3.11.0 or greater to be installed, and will not work without that.
That plugin is now included in phplist so you should only need to enable it.
See <https://github.com/bramley/phplist-plugin-common>

### Install through phplist ###
Install on the Plugins page (menu Config > Plugins) using the package URL `https://github.com/bramley/phplist-plugin-segment/archive/master.zip`

### Install manually ###
Download the plugin zip file from <https://github.com/bramley/phplist-plugin-segment/archive/master.zip>

Expand the zip file, then copy the contents of the plugins directory to your phplist plugins directory.
This should contain

* the file SegmentPlugin.php
* the directory SegmentPlugin

### Settings ###
In the Segmentation group on the Settings page you can specify:

* The size of the list of previous campaigns for Campaign activity. The default is 10.
* The number of subscribers to display who meet the segment conditions. The default is 50.

## Usage ##

For guidance on usage see the plugin page within the phplist documentation site <https://resources.phplist.com/plugin/segment>

## Support ##

Please raise any questions or problems in the user forum <https://discuss.phplist.org/>.

## Donation ##
This plugin is free but if you install and find it useful then a donation to support further development is greatly appreciated.

[![Donate](https://www.paypalobjects.com/en_US/i/btn/btn_donate_LG.gif)](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=W5GLX53WDM7T4)

## Version history ##

    version         Description
    2.14.1+20231212 Minor internal changes
    2.14.0+20231018 When viewing message show number of pending subscribers
    2.13.2+20230403 Update translations
    2.13.1+20230103 Fix problem with an invalid regex causing an exception and a "white page"
    2.13.0+20221104 Add IS NOT INCLUDED operator
    2.12.1+20210428 Allow the plugin to be a dependency of phplist
    2.12.0+20210225 Simplify whether campaign can be submitted
    2.11.0+20201212 Support aggregated campaigns for sent and clicked
    2.10.1+20200515 Make the dependency check message clearer
    2.10.0+20200307 Allow searching select list of subscriber fields and attributes
    2.9.0+20191231  Add aggregated campaigns to subscriber campaigns opened/not opened
    2.8.0+20190212  Add anniversary operator for date attribute field
    2.7.0+20181115  Add condition for date subscribed to list
    2.6.1+20181104  When sending a campaign log an event showing the number of subscribers selected
    2.6.0+20181103  Add condition for subscriber belonging to all lists selected for the campaign
    2.5.1+20181003  Fix earlier regression that limited the number of attributes to 15
    2.5.0+20180906  Segment according to a textarea filled with emails
    2.4.0+20180905  Add export of selected subscribers
    2.3.0+20180722  Display subscribers that meet the segment conditions
    2.2.7+20180517  Avoid dependency on php 5.6
    2.2.6+20180328  Reduce the level of php errors that are reported
    2.2.5+20171218  Improve the layout of the Segment tab
    2.2.4+20171109  Improve display of multiple-select
    2.2.3+20171023  Improvements to layout of the Segment tab
    2.2.2+20171018  Display warning when lists have not been selected
    2.2.1+20170208  Copy segment fields when copying a campaign
    2.2.0+20170126  Improvements to use of saved segments
    2.1.11+20161217 Extra validation of from and to dates
                    Use flatpickr for date input
    2.1.10+20160725 Calculate number of subscribers when there are not any conditions
    2.1.9+20160626  Fix problem whereby 0 was not accepted in text fields
    2.1.8+20160515  Fix sql error introduced in version 2.1.6
    2.1.7+20160515  Minor changes
    2.1.6+20160513  Fix for bug when using MESSAGEQUEUE_PREPARE
    2.1.5+20160316  Handle condition for an attribute that no longer exists
    2.1.4+20151117  Internal code refactoring
    2.1.3+20151024  Internal changes to meet coding standards
    2.1.2+20151019  Fix php warning
    2.1.1+20150912  Fix problem when viewing message
    2.1.0+20150903  Added "after interval" operator for date fields
                    Improve validation and warning of invalid conditions
    2.0.0+20150811  Display segment conditions on view message page
    2015-07-16      Can now save and re-use segments
    2015-06-02      Handle attribute not existing
    2015-05-06      Add dependency checks
    2015-04-04      Internal changes to improve memory usage
    2015-03-23      Change to autoload approach
    2015-03-04      Add subscriber id and uniqid as subscriber data fields
    2015-02-13      Add "is between" operator for date fields
    2015-01-06      Remove limit of 15 attributes
    2014-11-21      Allow text to be translated
    2014-10-20      Campaign activity select list is limited to prior campaigns sent to the lists
    2014-10-18      Fix bug in date attribute processing
    2014-10-15      Added list exclusion, internal changes
    2014-10-04      Add sent/not sent a campaign, and clicked/did not click any link
    2014-10-03      Add any/all, multi-value for select list and radio button attributes
    2014-09-27      Correct error reporting
    2014-09-26      Pull Request #1
    2014-09-25      Add regexp matching
    2014-09-24      Release to GitHub


changelog.txt
version 2.14.1+20231212
      Remove unnecessary parameter to queryOne()
      Use namespaced names for Common plugin classes instead of PEAR names

version 2.14.0+20231018
      When viewing message show number of pending subscribers
      cs fixer

version 2.13.2+20230403
      Update translations
      Update CS Fixer

version 2.13.1+20230103
      Fix problem with an invalid regex causing an exception and a "white page"

version 2.13.0+20221104
Aris Lydopoulos (1):
      Add IS NOT INCLUDED operator for selecting email addresses that are not in a list.

version 2.12.1+20210428
      Allow the plugin to be a dependency of phplist.

version 2.12.0+20210225
      Add base translations
      Simplify whether campaign can be submitted

version 2.11.0+20201212
      Correct docblock comments
      Support aggregated campaigns for sent and clicked

version 2.10.1+20200515
      Make the dependency check message clearer
      Bring PHP CS Fixer up to date

version 2.10.0+20200307
      Rework handling of attribute DAO
      Make the select list of subscriber fields and attributes searchable using Select2

version 2.9.0+20191231
      Add aggregated campaigns to subscriber campaigns opened/not opened

version 2.8.0+20190212
      Add anniversary operator for date attribute field
      Reinstate processing of IS operator for date fields

version 2.7.0+20181115
      Add condition for date subscribed to list
      Refactor joinQuery() for each date field
      Update cs config to not apply yoda rules.

version 2.6.1+20181104
      When sending a campaign log an event showing the number of subscribers selected
      Update English translations

version 2.6.0+20181103
      Add condition for subscriber belonging to all lists selected for the campaign
      Rework how messageData is provided when creating a condition
      Rename method and variables for condition type
      Fix error in earlier commit of creating segment only once

version 2.5.1+20181003
      Fix earlier regression that limited the number of attributes to 15.
      Fixes #18

version 2.5.0+20180906
Aris (1):
      Option to segment according to a textarea filled with emails

version 2.4.0+20180905
      Refactor condition handling into new class
      Move calculation of subscribers into the Segment class
      Store selected subscribers for each campaign
      Refactor use of limit when calculating subscribers
      Add export of selected subscribers
      Add dependency container
      Improve handling of empty and changed fields
      Ensure segment is created only once
      Further refactoring

version 2.3.0+20180722
      Simplify condition factory and use of attributes
      Display subscribers that meet the segment conditions

version 2.2.7+20180517
      Remove constant scalar expression to avoid dependency on php 5.6

version 2.2.6+20180328
      Reduce the level of php errors that are reported.
      Allow warning to be translated. Fixes #15.

version 2.2.5+20171218
      Improve the layout of the Segment tab

version 2.2.4+20171109
      Improve display of multiple-select
      Clarify the warning displayed when no lists have been selected. Fixes #13.
      Add English translation file as basis for other languages

version 2.2.3+20171023
      Display the saved segments fields only when there is at least one.
      Display the save segment input field and load button only when there is at least one condition.
Sam Tuke:
      Left align condition rows, other layout improvements. Fixes #12

version 2.2.2+20171018
      Add warnings when no lists have been selected or no campaigns sent to the selected lists.
      Re-position some of the command buttons.
      Refactor creating table aliases
      Coding standard changes

version 2.2.1+20170208
      Upgrade for version 2 of cs fixer. Use recommended file name.
      Coding standards changes, mostly remove trailing white space.
      Always display save segment fields but disable when there are no entered conditions.
      Hook for copying a campaign

version 2.2.0+20170126
      Correct query to work with ONLY_FULL_GROUP_BY.
      Multiselect when adding saved segments.
      Saved segment conditions are added to existing conditions instead of replacing them.
      Button to remove all conditions.
      Remove combine when saving/loading saved segment

version 2.1.11+20161217
      Validate that to date is not earlier than the from date.
      Use flatpickr package instead of jquery-ui for date input.

version 2.1.10+20160725
      Calculate number of subscribers when there are not any conditions

version 2.1.9+20160626
      Fix problem whereby 0 was not accepted in text fields

version 2.1.8+20160515
      Fix sql error introduced in version 2.1.6

version 2.1.7+20160515
      Use BitArray class to hold selected subscribers
      Add class map

version 2.1.6+20160513
      Include subscribers who have 'todo' status, set by MESSAGEQUEUE_PREPARE

version 2.1.5+20160316
      Handle condition for an attribute that no longer exists Fixes #9
      Coding standards changes
      Update copyright year

version 2.1.4+20151117
      Refactoring of id and uniqid

version 2.1.3+20151024
      Revised docblock
      Coding standards changes

version 2.1.2+20151019
      Add typehint

version 2.1.1+20150912
      Handle change of field when viewing

version 2.1.0+20150903
      Autosubmit form when operator is changed
      Display second field only for between operator
      Handle condition invalid by displaying warning or logging to event log
      New operator for date fields to send after interval

version 2.0.0+20150811
      Display segment conditions on view message page
      Add documentation url for plugin page

version 2015-07-16
      Add functionality to save and re-use segment conditions


plugins/SegmentPlugin.php
<?php
/**
 * SegmentPlugin for phplist.
 *
 * This file is a part of SegmentPlugin.
 *
 * SegmentPlugin is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * SegmentPlugin is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * @category  phplist
 *
 * @author    Duncan Cameron
 * @copyright 2014-2017 Duncan Cameron
 * @license   http://www.gnu.org/licenses/gpl.html GNU General Public License, Version 3
 */
use phpList\plugin\Common\Container;
use phpList\plugin\Common\ImageTag;
use phpList\plugin\Common\PageLink;
use phpList\plugin\Common\PageURL;
use phpList\plugin\SegmentPlugin\Segment;

class SegmentPlugin extends phplistPlugin
{
    const VERSION_FILE = 'version.txt';
    const GUIDANCE = 'https://resources.phplist.com/plugin/segment#add_segment_conditions';

    /*
     *  Inherited variables
     */
    public $name = 'Segmentation';
    public $authors = 'Duncan Cameron';
    public $description = 'Send to a subset of subscribers using custom conditions';
    public $documentationUrl = 'https://resources.phplist.com/plugin/segment';
    public $settings;

    private $error_level;
    private $selectedSubscribers = array();
    private $dao;
    private $conditionFactory;
    private $logger;
    private $segment = null;

    /**
     * Constructor.
     */
    public function __construct()
    {
        $this->coderoot = dirname(__FILE__) . '/' . __CLASS__ . '/';
        $this->settings = array(
            'segment_campaign_max' => array(
                'description' => s('The maximum number of earlier campaigns to select from'),
                'type' => 'integer',
                'value' => 10,
                'allowempty' => 0,
                'min' => 4,
                'max' => 25,
                'category' => 'Segmentation',
            ),
            'segment_subscribers_max' => array(
                'description' => s('The maximum number of selected subscribers to display'),
                'type' => 'integer',
                'value' => 50,
                'allowempty' => 0,
                'min' => 5,
                'max' => 500,
                'category' => 'Segmentation',
            ),
            'segment_saved_summary' => array(
                'description' => s('Summary of saved segments'),
                'type' => 'textarea',
                'value' => '',
                'allowempty' => true,
                'category' => 'Segmentation',
            ),
            'segment_attribute_max_length' => array(
                'description' => s('Limit the display length of an attribute name. Enter 0 to always display the full attribute name'),
                'type' => 'integer',
                'value' => 0,
                'allowempty' => true,
                'min' => 0,
                'max' => 100,
                'category' => 'Segmentation',
            ),
        );
        $this->error_level = E_ALL ^ E_NOTICE ^ E_DEPRECATED ^ E_STRICT;

        parent::__construct();
        $this->version = (is_file($f = $this->coderoot . self::VERSION_FILE))
            ? file_get_contents($f)
            : '';
    }

    /**
     * Provide the dependencies for enabling this plugin.
     *
     * @return array
     */
    public function dependencyCheck()
    {
        global $plugins;

        return array(
            'phpList version 3.3.2 or later' => version_compare(VERSION, '3.3.2') >= 0,
            'Common plugin version 3.11.0 or greater must be enabled' => (
                phpListPlugin::isEnabled('CommonPlugin')
                && version_compare($plugins['CommonPlugin']->version, '3.11.0') >= 0
            ),
            'PHP version 5.4.0 or greater' => version_compare(PHP_VERSION, '5.4') > 0,
        );
    }

    public function activate()
    {
        parent::activate();

        $depends = require $this->coderoot . 'depends.php';
        $container = new Container($depends);
        $this->dao = $container->get('SegmentPlugin_DAO');
        $this->conditionFactory = $container->get('ConditionFactory');
        $this->logger = $container->get('Logger');
    }

    /**
     * List of items to add to menu.
     *
     * @return array
     */
    public function adminmenu()
    {
        return array();
    }

    /**
     * The title of the additional tab.
     *
     * @param int $messageId
     *
     * @return string the title
     */
    public function sendMessageTabTitle($messageid = 0)
    {
        return s('Segment');
    }

    /**
     * Build contents of additional tab.
     *
     * @param int   $messageId
     * @param array $messageData
     *
     * @return string the html for the tab
     */
    public function sendMessageTab($messageId = 0, $messageData = array())
    {
        global $pagefooter;

        $er = error_reporting($this->error_level);

        $formFields = isset($messageData['segment']) ? $messageData['segment'] : array();
        $segment = $this->createSegment($messageId, $formFields);
        $this->processSegmentCommands($segment, $formFields);

        $selectPrompt = s('Select ...');
        $params = array();
        $params['condition'] = array();
        $params['selectPrompt'] = $selectPrompt;

        foreach ($segment->conditions() as $i => $c) {
            $s = new stdClass();
            $params['condition'][] = $s;
            $field = $c['field'];

            try {
                $type = $this->conditionFactory->createConditionType($field, $messageData);
            } catch (SegmentPlugin_ConditionException $e) {
                $s->error = s('Unable to create condition - %s', $e->getMessage());
                continue;
            }

            // display field selection drop-down list
            $s->fieldList = $this->fieldDropDownList($field, $i, $selectPrompt);

            // display hidden input to detect when field changes
            $s->hiddenField = CHtml::hiddenField("segment[c][$i][_field]", $field);

            // display operators drop-down list
            $operators = $type->operators();
            $selected = isset($c['op']) ? $c['op'] : key($operators);
            $s->operatorList = CHtml::dropDownList(
                "segment[c][$i][op]",
                $selected,
                $operators,
                array('class' => 'autosubmit')
            );

            // display value
            $value = isset($c['value']) ? $c['value'] : '';
            $s->display = $type->display($selected, $value, "segment[c][$i]");
        }

        // add empty field
        $s = new stdClass();
        $i = count($segment->conditions());
        $s->fieldList = $this->fieldDropDownList('', $i, $selectPrompt);
        $s->hiddenField = CHtml::hiddenField("segment[c][$i][_field]", '');
        $params['condition'][] = $s;

        // display warning if no lists have been selected
        if (!(is_array($messageData['targetlist']) && count($messageData['targetlist']) > 0)) {
            $params['warning'] = s('Please select at least one list on the Lists tab before adding segmentation conditions');
        }

        // display fields for saved segments only where there are some
        $saved = new SegmentPlugin_SavedSegments();
        $savedListData = $saved->selectListData();

        if (count($savedListData) > 0) {
            $params['savedList'] = CHtml::dropDownList(
                'segment[usesaved][]',
                '',
                $savedListData,
                array('multiple' => 'multiple', 'style' => 'width: 50%')
            );
            $params['loadButton'] = CHtml::submitButton(s('Load segments'), array('name' => 'segment[load]'));
            $params['settingsButton'] = new PageLink(
                new PageURL('configure', array(), 'segmentation'),
                'Edit saved segments',
                array('target' => '_blank', 'class' => 'button')
            );
        }

        // display combine drop-down list
        $params['combineList'] = CHtml::dropDownList(
            'segment[combine]',
            $segment->combine(),
            array(SegmentPlugin_Operator::ONE => s('any'), SegmentPlugin_Operator::ALL => s('all'))
        );

        // display calculate button
        $params['calculateButton'] = CHtml::submitButton(s('Calculate'), array('name' => 'segment[calculate]'));

        // display calculated number of subscribers
        if (isset($formFields['calculate'])) {
            try {
                list($params['totalSubscribers'], $params['subscribers']) = $segment->calculateSubscribers(getConfig('segment_subscribers_max'));
                $params['exportCalculatedButton'] = new PageLink(
                    new PageURL('export', array('pi' => 'SegmentPlugin', 'id' => $messageId)),
                    'Export subscribers',
                    array('class' => 'button dialog')
                );
            } catch (SegmentPlugin_ValueException $e) {
                $params['warning'] = s('Invalid value for segment condition');
            } catch (SegmentPlugin_ConditionException $e) {
                $params['warning'] = s('Unable to create condition - %s', $e->getMessage());
            }
        }

        // display remove all, save button and input field only when there is at least one entered condition
        if (count($segment->conditions()) > 0) {
            $params['removeButton'] = CHtml::submitButton(s('Remove all'), array('name' => 'segment[remove]'));
            $params['saveButton'] = CHtml::submitButton(s('Save segment'), array('name' => 'segment[save]'));
            $params['saveName'] = CHtml::textField('segment[savename]', '', array('size' => 25, 'placeholder' => 'Name of segment'));
        }

        // display link to Help page
        $params['help'] = CHtml::tag(
            'a',
            array('href' => self::GUIDANCE, 'target' => '_blank'),
            new ImageTag('info.png', 'Guidance')
        );
        $html = $this->render('sendtab.tpl.php', $params);
        $pagefooter[basename(__FILE__)] = file_get_contents($this->coderoot . 'script.html');
        error_reporting($er);

        return $html;
    }

    /**
     * Validate that conditions are valid for the message to be submitted by calculating the number of subscribers.
     *
     * @param array $messageData
     *
     * @return string empty string for allow, otherwise error text to be displayed
     */
    public function allowMessageToBeQueued($messageData = array())
    {
        if (!isset($messageData['segment']['c'])) {
            return '';
        }
        $segment = $this->createSegment($messageData['id'], $messageData['segment']);

        try {
            $segment->validateSegment();
        } catch (SegmentPlugin_ValueException $e) {
            return s('Invalid value for segment condition');
        } catch (SegmentPlugin_ConditionException $e) {
            return s('Unable to create condition - %s', $e->getMessage());
        }

        return '';
    }

    /**
     * Use this hook to delete the 'not sent' rows from the usermessage table
     * so that they will be re-evaluated.
     *
     * @param int $id the message id
     *
     * @return none
     */
    public function messageQueued($id)
    {
        $this->dao->deleteNotSent($id);
    }

    /**
     * The same processing as when queueing a message.
     *
     * @param int $id the message id
     *
     * @return none
     */
    public function messageReQueued($id)
    {
        $this->messageQueued($id);
    }

    /**
     * Use this hook to select the subscribers who meet the segment conditions.
     * $selectedSubscribers will contain the selected subscribers.
     *
     * @param array $messageData the message data
     *
     * @return none
     */
    public function campaignStarted($messageData = array())
    {
        $er = error_reporting($this->error_level);
        $messageId = $messageData['id'];
        $this->selectedSubscribers[$messageId] = null;

        if (isset($messageData['segment']['c'])) {
            $segment = $this->createSegment($messageId, $messageData['segment']);

            try {
                list($total, $this->selectedSubscribers[$messageId]) = $segment->loadSubscribers();
                logEvent(s('Segment plugin selected %d subscribers for campaign %d', $total, $messageId));
            } catch (SegmentPlugin_ValueException $e) {
                logEvent(s('Invalid value for segment condition'));
            } catch (SegmentPlugin_ConditionException $e) {
                logEvent(s('Unable to create condition - %s', $e->getMessage()));
            } catch (SegmentPlugin_NoConditionsException $ex) {
                // do nothing
            }
        }
        error_reporting($er);
    }

    /**
     * Determine whether the campaign should be sent to a specific user.
     *
     * @param array $messageData the message data
     * @param array $userData    the user data
     *
     * @return bool
     */
    public function canSend($messageData, $userData)
    {
        $messageId = $messageData['id'];
        $userId = (int) $userData['id'];

        return ($this->selectedSubscribers[$messageId] === null)
            ? true
            : (bool) $this->selectedSubscribers[$messageId][$userId];
    }

    /**
     * Build the html to be added to the view message page.
     *
     * @param int   $messageId   the message id
     * @param array $messageData the message data
     *
     * @return array|false the caption and html to be added, or false if the message
     *                     does not use segments
     */
    public function viewMessage($messageId, array $messageData)
    {
        if (!isset($messageData['segment'])) {
            return false;
        }
        $formFields = $messageData['segment'];

        if (!isset($formFields['c'])) {
            return false;
        }
        $segment = $this->createSegment($messageId, $formFields);
        $conditions = $segment->conditions();

        if (count($conditions) == 0) {
            return false;
        }
        $params = array();
        $params['condition'] = array();

        foreach ($conditions as $i => $c) {
            $s = new stdClass();
            $params['condition'][] = $s;
            $field = $c['field'];

            try {
                $type = $this->conditionFactory->createConditionType($field, $messageData);
            } catch (SegmentPlugin_ConditionException $e) {
                $s->error = sprintf('Unable to create condition - %s', $e->getMessage());
                continue;
            }

            // display field selection
            $fields = $this->conditionFactory->subscriberFields() + $this->conditionFactory->attributeFields();
            $s->field = $fields[$field];

            // display operator
            $operators = $type->operators();
            $op = isset($c['op']) ? $c['op'] : key($operators);
            $s->operator = $operators[$op];

            // display value field
            $value = isset($c['value']) ? $c['value'] : '';
            $s->display = $type->display($op, $value, "segment[c][$i]");
        }

        // display combine
        $combineOps = array(SegmentPlugin_Operator::ONE => s('any'), SegmentPlugin_Operator::ALL => s('all'));
        $params['combine'] = $combineOps[$segment->combine()];
        list($params['totalSubscribers']) = $segment->calculateSubscribers(1);
        $html = $this->render('viewmessage.tpl.php', $params);

        return array('Segment conditions', $html);
    }

    /**
     * Called when a campaign is being copied.
     * Allows this plugin to specify which rows of the messagedata table should also
     * be copied.
     *
     * @return array rows of messagedata table that should be copied
     */
    public function copyCampaignHook()
    {
        return array('segment');
    }

    private function render($template, $params)
    {
        extract($params);
        ob_start();
        require $this->coderoot . $template;

        return ob_get_clean();
    }

    private function fieldDropDownList($field, $seq, $selectPrompt)
    {
        return CHtml::dropDownList(
            "segment[c][$seq][field]",
            $field,
            array('Subscriber Data' => $this->conditionFactory->subscriberFields(), 'Attributes' => $this->conditionFactory->attributeFields()),
            array('prompt' => $selectPrompt, 'class' => 'autosubmit searchable')
        );
    }

    /**
     * Process segment commands that can change the set of conditions.
     *
     * @param Segment $segment
     * @param array   $formFields
     */
    private function processSegmentCommands($segment, $formFields)
    {
        $saved = new SegmentPlugin_SavedSegments();

        if (isset($formFields['remove'])) {
            /*
             *  Remove all conditions
             */
            $segment->removeAll();
        } elseif (isset($formFields['save']) && $formFields['savename'] != '') {
            /*
             *  Save the current set of conditions
             */
            $saved->addSegment($formFields['savename'], $segment->conditions());
        } elseif (isset($formFields['load']) && isset($formFields['usesaved']) && is_array($formFields['usesaved'])) {
            /*
             *  Load the selected saved segments
             */
            try {
                foreach ($formFields['usesaved'] as $savedId) {
                    $savedConditions = $saved->segmentById($savedId);
                    $segment->addConditions($savedConditions);
                }
            } catch (Exception $e) {
                // do nothing
            }
        }
    }

    /**
     * Create a segment from the message form fields.
     *
     * @param int   $messageId
     * @param array $formFields
     *
     * @return Segment
     */
    private function createSegment($messageId, $formFields)
    {
        return new Segment(
            $messageId,
            isset($formFields['c']) ? $formFields['c'] : array(),
            isset($formFields['combine']) ? $formFields['combine'] : SegmentPlugin_Operator::ALL,
            $this->conditionFactory
        );
    }
}
